xwiki-pro-macros-api/src/main/java/com/xwiki/macros/internal/updateReferences/listeners/PageRenameMacroParametersUpdateListener.java

```diff
@@ -0,0 +1,164 @@
+import org.xwiki.job.JobContext;
+import org.xwiki.job.event.JobStartedEvent;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.ObservationContext;
+import org.xwiki.observation.event.Event;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xwiki.macros.internal.updateReferences.UpdateReferencesQueueEntry;
+import com.xwiki.macros.internal.updateReferences.UpdateReferencesRunnable;
+
+/**
+ * Listens to rename of pages. Also, for pages that contains macros with references it will start a thread for updating
+ * references of the macros parameters that are a document reference.
+ *
+ * @version $Id$
+ * @since 1.29
+ */
+@Component
+@Named(PageRenameMacroParametersUpdateListener.ROLE_HINT)
+@Singleton
+public class PageRenameMacroParametersUpdateListener extends AbstractEventListener implements Disposable
+{
+    protected static final String ROLE_HINT = "PageRenameMacroParametersUpdateListener";
+
+    private static final Map<String, List<String>> MACROS_TO_UPDATE = Map.of(
+        "button", List.of("url"),
+        "confluence_pagetree", List.of("root")
+        );
+
+    /**
+     * Thread that will handle updating the reference of a macro parameter after a page rename.
+     */
+    public Thread referencesMacroThread;
+
+    @Inject
+    protected JobContext jobContext;
+
+    @Inject
+    protected ObservationContext observationContext;
+
+    @Inject
+    private UpdateReferencesRunnable referencesMacroRunnable;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Constructor.
+     */
+    public PageRenameMacroParametersUpdateListener()
+    {
+        super(ROLE_HINT, new DocumentCreatedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        if (this.referencesMacroThread == null) {
+            startThreads();
+        }
+
+        if (observationContext.isIn(new JobStartedEvent("refactoring/rename"))) {
+            Job job = jobContext.getCurrentJob();
+            DocumentReference destinationRef = job.getRequest().getProperty("destination");
+            List<DocumentReference> references = job.getRequest().getProperty("entityReferences");
+
+            if (references != null && !references.isEmpty()) {
+                XWikiDocument currentDoc = (XWikiDocument) source;
+
+                DocumentReference originalDocRef = references.get(0);
+                DocumentReference currentDocRef = currentDoc.getDocumentReference();
+
+                if (destinationRef.equals(currentDocRef)) {
+                    startUpdatingReferences((XWikiContext) data, currentDoc, originalDocRef);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void dispose()
+    {
+        try {
+            stopThread(this.referencesMacroThread, this.referencesMacroRunnable);
```

Comment:
Same issue in the diagram application, but it seems to me we are starting a thread that will be sitting there forever (until the component is updated) as soon as a document is renamed.

Isn't this an issue? Shouldn't we make the thread leave when it doesn't have anything to do, maybe after some time?

Naively, I would think we could drop this heavy thread handling and use something like [CompletableFuture.runAsync](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#runAsync-java.lang.Runnable-)

cc @oanalavinia